### ISSUE TITLE
bucket.objects[object].exists? returns AWS::S3::Errors::Forbidden when object doesn't exist

### ISSUE THREAD

**Issue by cwholt** [aws/aws-sdk-ruby#201]
should return false.

**Comment by dmagliola** [comment]
Another "vote" for @shipstar's comment, for anyone reading this, that's definitely the problem.

**Comment by frankwallis** [comment]
works for me too, thanks @shipstar!

**Comment by vickeryj** [comment]
I had this problem when I specified an s3 path in my IAM policy (`arn:aws:s3:::bucket/path*`). Changing the policy per @shipstar's suggestion to grant access to the entire bucket fixed this. What's a good way to bring the up with the IAM people?

**Comment by kgf** [comment]
Does anyone know what permissions are actually needed ? Would prefer to not just wildcard...
What I currently have, but not enough yet.

`            "Effect": "Allow",
            "Action": [
                "s3:GetBucketLocation",
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::bucket-us-east-1"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:AbortMultipartUpload",
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::bucket-us-east-1/*"
            ]
        },`

**Comment by kgf** [comment]
Found this link, testing now
https://groups.google.com/forum/#!topic/fluentd/Vvok7FTVuq4

**Comment by coreystinson2** [comment]
tl;dr Adding `s3:ListBucket` instead of `s3:*` worked for me.

So this had been an ongoing issue with me for quite some time.

I have a library responsible for uploading new objects to s3, you can see the code below. I generate bucket keys at the time of upload however first do a quick existence check to prevent any accidental overrides if we **somehow** generate the same hex
```ruby
key = "#{SecureRandom.uuid}#{extension}"
existing = Aws::S3::Object.new(bucket, key)
if existing.exists?
  raise ExistsError, 'Please try uploading the file again.'
end
```

Now developing locally I use the following IAM permissions, I like keeping these down to the bare minimum for both my development and production policies.
```json
"Action": [
    "s3:GetObject",
    "s3:PutObject"
],
"Resource": [
    "arn:aws:s3:::my-bucket/*",
    "arn:aws:s3:::my-bucket"
]
```
The above policy works great, the upload library executes without problems and returns both true and false based on whether or not the file actually exists.

Now we push this code to production on an instance registered to ECS and we start experiencing some weird behaviour. This exact same policy now starts throwing the following error
`/usr/local/bundle/gems/aws-sdk-resources-2.5.3/lib/aws-sdk-resources/resource.rb:134:in rescue in exists?' `

While the aforementioned fixes such as changing the policy actions to below, do in fact work - I'm super anal about giving my applications excessive permissions and think it's a great practise to ensure **they only have what they need to have**.
```json
"Action": [
    "s3:*"
],
"Resource": [
    "arn:aws:s3:::my-bucket/*",
    "arn:aws:s3:::my-bucket"
]
```

Anyway I spent ages going one by one through all the policies until I found the combination that worked. `s3:ListBucket`.
```json
"Action": [
    "s3:GetObject",
    "s3:ListBucket",
    "s3:PutObject"
],
"Resource": [
    "arn:aws:s3:::my-bucket/*",
    "arn:aws:s3:::my-bucket"
]
```

Still havn't figured out the difference between development and production in regards to exactly why production requires this new action. However I'm much happier using this as oppose to just straight up `s3:*`

**Comment by AlexMorreale** [comment]
Just ran into this myself. It's pretty crazy `.exists?` requires `GetObject` and not just `ListBucket`

**Comment by jacksonrayhamilton** [comment]
I ran into this too, but it was not even a real permission problem, my server’s clock was off.  Using `ntp` fixed it for me: https://stackoverflow.com/a/49260424/1468130  Leaving this here in case it helps anyone else.

**Comment by pkoch** [comment]
Supposedly, just `s3:ListObjects` and `s3:GetObject` is sufficient. Source: https://aws.amazon.com/premiumsupport/knowledge-center/s3-troubleshoot-403/
